_pytest/python_api.py

```diff
@@ -443,8 +443,9 @@ def approx(expected, rel=None, abs=None, nan_ok=False):
     # This has the advantage that it made it easy to support mapping types
     # (i.e. dict).  The old code accepted mapping types, but would only compare
     # their keys, which is probably not what most people would expect.
-
-    if _is_numpy_array(expected):
+    if isinstance(expected, String):
```

Comment:
Perhaps it would be better if we change the `else:` clause here to check for `float`/`int`, and then add an `else:` clause which fails with everything else?

```python
    elif isinstance(expected, Decimal):
        cls = ApproxDecimal
    elif isinstance(expected, (float,) + six.integer_types):
        cls = ApproxScalar
    else:
        raise TypeError('Unknown type passed to approx: {!r} ({!r})'.format(expected, type(expected))
```

Reply:
But isn't string the only thing that cannot currently be approx'd?

Reply:
Nevermind! I did some tests and noticed it doesnt accept things I thought it could. 
```

    if _is_numpy_array(expected):
        cls = ApproxNumpy
    elif isinstance(expected, Mapping):
        cls = ApproxMapping
    elif isinstance(expected, Sequence) and not isinstance(expected, String):
        cls = ApproxSequence
    elif isinstance(expected, Decimal):
        cls = ApproxDecimal
    elif isinstance(expected, (float,) + six.integer_types):
        cls = ApproxScalar
    else:
       raise TypeError('Unknown type passed to approx: {!r} ({!r})'.format(expected, type(expected))
```
Would be the final commit correct?

Im not sure how that gets us from passing in [6, (5,4)] though. Because the above clauses would see it as valid, but math.isinf freaks out about it (rightly so).

Reply:
Hey @Drew-Ack, thanks for replying so quickly!

> Im not sure how that gets us from passing in [6, (5,4)] though. Because the above clauses would see it as valid, but math.isinf freaks out about it (rightly so).

I'm not sure I follow, could you clarify please?

Reply:
No problem! Love pytest afterall :)

As an example this elif:
```
    elif isinstance(expected, Sequence) and not isinstance(expected, String):
        cls = ApproxSequence
```
allows this as valid [5, (6,7)] input into the approx function. 
I assume this is because a sequence like [5, 6, 7] is a valid sequence. 

The issue comes when math.isinf runs. math.isinf is evaluated in the ApproxSequence class. I show this in #3473

I assume this clause is used purely to delegate which class to use, but provides no checking to make sure the sequence is valid. 
Is this something we should try to do?
Should we error out when a sequence of [5, (6,7)] is given

Reply:
> allows this [5, (6,7)] as valid input into the approx function.

Oh my bad, it was obvious you meant `[5, (6, 7)]` the literal. Sorry about that.

> This technically shouldnt be, because even though its a sequence it will just blow up when appox runs.

I see. I guess we can iterate over every element of `Sequence` and `Mapping` to ensure their elements are `float` or `int` objects; there might be a performance problem but I don't think people are using `pytest.approx` with arrays so large that this would make a difference.

Reply:
This is the first open source project I've contributed too, so I will defer to you developers on what is best practice. 

It seems to me we are just picking what point to throw an exception at. 
I put together a small test case for it actually :)
```
def test_time():
    mylist = []
    for x in range(1000000):
        mylist.append(x)

    start_time = time.time()

    for x in mylist:
        if isinstance(x, float) or isinstance(x, int):
            continue

    end_time = time.time()

    print(end_time - start_time)
```

Its roughly .28 seconds to run through 1 million elements. 
Its roughly 2.8 seconds to run through 10 million.
The pattern holds as you increase by x10

Reply:
> This is the first open source project I've contributed too, so I will defer to you developers on what is best practice.

This is great, thanks for contributing to pytest, we definitely appreciate it.

What do you think @RonnyPfannschmidt? I think we can get away by keeping things simple and just going with my suggestion in https://github.com/pytest-dev/pytest/pull/3474#discussion_r188440030, but I don't oppose if others prefer to be more through.

Reply:
basically massive sequences shouldn't use python lists ^^ so the check should be reasonably fine for normal sequences

Reply:
OK, @Drew-Ack could go with a more through check for sequences then? Thanks!